### DISCUSSION TITLE
Running headless code, rendering markup.

### DISCUSSION THREAD

**Discussion by c4lliope** [taybenlor/runno#199] [General]
Hello @taybenlor and all,

I'm really eager to begin using Runno;
seems like a dead-simple solucion for a common, complex series of problems.

In my use case, I need some Ruby running in my pages,
only I'm hoping I can keep all my code processing in the background.
I'm planning on using Ruby for rendering some markup code,
and then embedding the markup my Ruby code produces in my page.

I'm only one day in on this goal,
and already I can see really good progress:

```javascript
import { RunnoProvider } from '@runno/runtime'

var processor = new RunnoProvider(
  null, // terminal
  null, // editor
  null, // controls
)

processor.headlessRunCode(
  `ruby`,
  `p 'Running Ruby...'`,
  '', // stdin
).then(response => {
  console.log(response)
})
```

This responds:
![2022-04-26_runno_response](https://user-images.githubusercontent.com/30454698/165341894-47b51742-7ccf-4073-8173-2ebd4b733183.png)

As you can see, the `runno` Ruby processor combines my code and the response
inside the logged `stdout` and `tty` variables.

I'm unsure the cause, clearly; perhaps this arrangement is necessary for running the terminal emulation.
Sadly, this breaks my hope for using `runno`-managed Ruby inside my pages,
unless I can begin relying on the `filesystem` approach for recording my rendered markup from Ruby.
I'm curious to read your opinions on this edge case of your library, @taybenlor !

I'm eager for a day when people can choose any language for rendering their single-page applications;
perhaps you can lead us in accomplishing this goal!

**Comment by taybenlor** [comment]
Hey Grace, this is definitely on the niche end of use cases!

Runno is definitely aimed at the use case of code examples / code snippets. I think you might be interested in two different directions:

1. Opal Ruby (Ruby compiled to JavaScript) - https://opalrb.com/ - this is more aimed at application development using Ruby in the web browser. The actual running code is JS so it will have execution differences to CRuby.

2. Ruby Wasm (Ruby running in Web Assembly) - https://github.com/ruby/ruby.wasm - There's a bunch of interesting stuff being done to run the full ruby runtime in the browser (this has only just happened in the last few months). There's an NPM package that you can use which is simpler than Runno and does basically exactly what you want.

I think the Ruby Wasm stuff is the most promising for what I think you want to do. It will be slower than Opal Ruby because it's running the full ruby runtime inside the browser, but that's what Runno does anyway.

**Reply by c4lliope** [reply]
Ah, I'm glad I can read up on Opal - speed is necessary in my case, and based on simple benchmarks,
seems like I need as much speed as possible in my `.erb` (Embedded RuBy language) markup processor.

My main concern in any library like these is the chances of adding language-specific packages, like Ruby's gems,
so you can merge in as much of a language's libraries as necessary. I hope more people can help compile and share packages.

**Comment by c4lliope** [comment]
Ended up running `ruby.wasm`, as you can see here: [://ruby.assembled.app](https://ruby.assembled.app) -
an upcoming challenge includes compiling [rubygems and c-based libraries](https://github.com/ruby/ruby.wasm/issues/14
) inside my packaged ruby binary.

**Comment by taybenlor** [comment]
Awesome, glad I could help! Looks very interesting :)
